### PR TITLE
Fixes #30445 - Host registration template & Insights with Rex setup

### PR DESCRIPTION
**Pull request by stejskalleos**
**Features:**
* Get & render _Host registration template_ in _Global registration template_
* New global parameters
```
host_registration_ssh_keys - default: true
host_registration_insights - default: false
```

**Limitations**
Before registration:
* _Host Registration template_ must have assigned OS
* OS must have assigned _Host Registration template_
* Hosts without OS (registered via `create_host_curl`) cannot render _Host registration template_

PR is part of **Simple & automatic host registration WF**, see:
* [RFC](https://community.theforeman.org/t/rfc-simple-automatic-host-registration-wf/19588)
* [Redmine (this PR)](https://projects.theforeman.org/issues/30445)
* [Redmine (Parent task)](https://projects.theforeman.org/issues/30440)

### CODE REVIEW

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 1]
```diff
@@ -44,13 +44,19 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
...
 <% end -%>
+
+if ! [ $? = 0 ]; then
+    echo ""
+    echo "Host registration failed"
+    exit 1
+fi
+
+TOKEN=$(curl -s -X GET "<%= foreman_server_url %>/api/hosts?search=name+%3D+$(hostname --fqdn)" <%= headers.join(' ') %> | grep -Po '(?<=\"registration_token\":")([^\"]+)')
+curl -X GET "<%= foreman_server_url %>/foreman_register/hosts/register?token=$TOKEN" | bash
```
Comment (timogoebel) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: Why do we double `curl` here?

cc: @ekohl

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 2]
```diff
@@ -44,13 +44,19 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
 
     curl --insecure --output $CONSUMER_RPM <%= subscription_manager_configuration_url %>
     yum localinstall $CONSUMER_RPM -y
-    yum install subscription-manager -y
     subscription-manager register --org=<%= @organization.label %> --activationkey="sample-key"
-
-    rm -f $CONSUMER_RPM
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb:50]: Why do you not want to clean this up?

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 3]
```diff
@@ -44,13 +44,19 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
 
     curl --insecure --output $CONSUMER_RPM <%= subscription_manager_configuration_url %>
     yum localinstall $CONSUMER_RPM -y
-    yum install subscription-manager -y
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb:47]: This is showing up in multiple PRs now. I think it can be its own PR and easily merged. You can also drop `--insecure` since the URL is HTTP.

File: app/views/unattended/provisioning_templates/registration/linux_registration_default.erb [hunk 4]
```diff
@@ -20,6 +20,19 @@ set -e
 <%= snippet 'puppet_setup' %>
 <% end -%>
 
+<% if host_param_true?('host_registration_insights') %>
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/linux_registration_default.erb]: Please remember line endings and use `-%>`

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 2 above)
Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:50]: When host is created, I want to check if everything went well or exit the script. Because we use two methods for host registration (`create_host_curl` &  `subscription-manager`), the easiest way was to check status of the last command. See:
```
if ! [ $? = 0 ]; then
    echo ""
    echo "Host registration failed"
    exit 1
fi
```
It worked well for `create_host_curl`, but not for the case with katello & rhel/centos, where last command is `rm -f $CONSUMER_RPM`.

And since the size of file is only `8.6K` and it's created in `/tmp` folder (where it will be deleted after a while, right?), I thought that keeping the file there won't be a problem.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 1 above)
Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: 1. Get token for Host registration template
2. Use the `$TOKEN` for rendering the Host registration template

Comment (timogoebel) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: Why do we need an explicit host registration template? I think one call to Foreman for registering a host should be enough.
You could say, I'm nitpicking here, but how would you explain to a user what the global registration template does and what the host-specific template does.

Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: > Why do we need an explicit host registration template? I think one call to Foreman for registering a host should be enough.

Yes, that is true but we want to also do some setup on the host. In this case we want to setup insights & remote execution. 
And for this _Host registration template_ is ideal.

In _Host registration template_ we have access to the `@host` object, so users can adjust registration behavior for specific OS or host group by configuring host parameters.

For example, you want to install Insights on the RHEL & CentOS but not on Debian and others,
or setup different ssh keys for different OS.

> You could say, I'm nitpicking here, but how would you explain to a user what the global registration template does and what the host-specific template does.

You are right, the naming can be little bit confusing, but it's pretty easy:
**Global Registration template** - Register unknown host to the Foreman.
**Host registration template** - Setup the host (Install puppet, insights, setup REX ...)

Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: Maybe it **Host setup template** would be better than **Host registration template**

Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: I think that calling register again is a bad design. If you want such a thing, I'd recommend implementing a special action to create the host. You currently do a `POST` to the hosts API, but what if you implement `POST /register` to create the host? That would then return a rendered template with additional actions to be performed. This avoids a request and makes authentication easier. It would also be easier on a Smart Proxy because you only need to deal with a single endpoint.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 2 above)
Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:50]: I'd suggest to avoid making the whole script idempotent. That's hard and I wouldn't want to put that burden on users who want to modify this. See https://github.com/theforeman/foreman/pull/7976#discussion_r486968427 for my suggestion to avoid needing this.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 1 above)
Comment (ares) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: how would you authenticate the `POST /register`? note that we get the registration token in the first curl, because we can't generate it before the host exists. And it exists only after we do the POST request.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 5]
```diff
@@ -46,11 +46,18 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
...
 create_host_curl
 <% end -%>
+
+if ! [ $? = 0 ]; then
+    echo ""
+    echo "Host registration failed"
+    exit 1
+fi
+
+TOKEN=$(curl -s -X GET "<%= foreman_server_url %>/api/hosts?search=name+%3D+$(hostname --fqdn)" <%= headers.join(' ') %> | grep -Po '(?<=\"registration_token\":")([^\"]+)')
```
Comment (ares) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: This detail call shouldn't be necessary, we can read all of that from host creation by parsing the json output we get from it.

This won't work for the sub-man path, so we need to do this only in that branch. Also it's not necessary to use search here, this would also work

```
"<%= foreman_server_url %>/api/hosts/$(hostname --fqdn)/"
```

Comment (ares) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: Second comment, we shouldn't be hardcoding paths in templates (/api/hosts). Instead we should compute this value in the controller that renders the template and pass it as a `@variable`. So we'd end up with something like `curl -s -X GET @registration_url | bash`

File: app/views/unattended/provisioning_templates/registration/linux_registration_default.erb [hunk 6]
```diff
@@ -20,6 +20,19 @@ set -e
 <%= snippet 'puppet_setup' %>
 <% end -%>
 
+<% if host_param_true?('host_registration_insights') -%>
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/linux_registration_default.erb]: This also needs to check if it's RHEL since insights is only available there.

File: app/views/unattended/provisioning_templates/registration/linux_registration_default.erb [hunk 7]
```diff
@@ -20,6 +20,19 @@ set -e
...
+echo '#'
+echo '# Installing Insights client'
+echo '#'
+
+yum install insights-client
+insights-client --register
+<% end -%>
+
+<% if host_param_true?('host_registration_rex') -%>
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/linux_registration_default.erb]: I'd drop this parameter since all others also call it unconditionally:
https://github.com/theforeman/foreman/blob/fc2387648abc558d8f3a8d9f4d914530f5823587/app/views/unattended/provisioning_templates/finish/kickstart_default_finish.erb#L58-L60

If we want a parameter to control that, it should behave the same for all methods (provisioning, registration, whatever). That is out of scope for this PR so I'd leave it out.

File: app/views/unattended/provisioning_templates/registration/linux_registration_default.erb [hunk 8]
```diff
@@ -20,6 +20,19 @@ set -e
 <%= snippet 'puppet_setup' %>
 <% end -%>
 
+<% if host_param_true?('host_registration_insights') -%>
+echo '#'
+echo '# Installing Insights client'
+echo '#'
+
+yum install insights-client
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/linux_registration_default.erb]: I think you missed a `-y` here

File: app/views/unattended/provisioning_templates/registration/linux_registration_default.erb [hunk 9]
```diff
@@ -20,6 +20,19 @@ set -e
...
+insights-client --register
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/linux_registration_default.erb]: Be aware that if this fails for whatever reason, the script execution halts here (due to `set -e` earlier). That means that REX is not set up either so it's not easy to recover. That's why I would set up REX before insights.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 1 above)
Comment (ares) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: We spent some time on this today with @ekohl. This would work in vanilla Foreman scenario, but when subscription-manager is involved, that would be a different flow. subscription-manager is unable to execute the template it would receive and even if we implement it, it wouldn't be backported to old RHELs. We'll eliminate the need for fetching details, but we'll still need to get the `Host registration template` at the end of the `Global registration template`.

Hope this makes it easier to visualize in mind, the documentation is not available yet, but will be once the code is in... Global template is generic recipe, how to register to this Foreman instance. As part of that it preps the system (e.g. installs the cert in order to securely talk to Foreman) and creates the initial record in Foreman DB. Then it tell host to start bootstrapping. That is handled by Host registration template, a specific template for a given host. That may differ a lot based on chosen hostgroup, OS or other trivia we know about that host. That's why it's specific to the host. Getting this template switches the host into "started registration" build mode. After it's done, it call back home informing Foreman it's built (same as in case of kickstart/preseed). Thanks to that, user will know when host is ready to use.

File: app/views/unattended/provisioning_templates/registration/linux_registration_default.erb (hunk 7 above)
Comment (ares) [inline app/views/unattended/provisioning_templates/registration/linux_registration_default.erb]: As discussed today, we don't need to wrap this in parameter, kickstart also does not do it. The snippet already checks if the param is set. In some later PR, we'll just change the logic a bit to seed the param but default to false. But for this PR, let's not add it yet. (let's keep it for insights though, that is a new thing)

File: db/seeds.d/180-common_parameters.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+CommonParameter.without_auditing do
+  params = [
+    { name: "host_registration_rex", key_type: "boolean", value: true },
```
Comment (ares) [inline db/seeds.d/180-common_parameters.rb]: we won't need this one yet

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 2 above)
Comment (ares) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:50]: we could save the result of curl/sub-man call to a variable and check that instead of $?

**Review by ares — CHANGES_REQUESTED** [review]
several comments left after discussion with Ewoud

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 2 above)
Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:50]: Looks like I didn't submit my comment, but the easier solution is to first `rm -f $CONSUMER` and then run `subscription-manager`. Then you still have `$?`. That's a good thing regardless, but storing is a good option as well.

Another common pattern for this is:

```sh
die() {
  echo "$*"
  exit 1
}

my_command || die "Failed to do x"
```

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 1 above)
Comment (timogoebel) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: To be completely honest, I don't get what you're trying to say here. Maybe it's because I don't fully understand the subscription-manager based workflow, but I think we should design this in a way where it does not really matter what particular tools are used to manage the host.

Let's start with a simple question:
In the current implementation of the global registration template, why is there a basically unauthorized call to `subscription-manager`? Doesn't this belong in the host-specific registration template that has all the required information about the host, possibly a one-time activation key tailored for that system?
Can't we keep this simple and just create the host in the global registration template and handle the rest in the host-specific registration template?

Maybe having a picture visualizing the workflow would help.
Let us clarify this please, before we proceed here.

Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: My understanding is that there are roughly 2 ways that we support to create the Host object:
* Via API
* Having subscription-manager upload facts

Since the API is clear to you, I won't get into that. I'll also note that my understanding of subscription-manager is only based on what others have told me.

Subscription-manager is used with an activation key. That key should be treated as a secret (even though often it isn't) since it is used to authenticate the request to create a host object and issue candlepin certificates. Subman also uploads facts. It replaces the API call since it does more. Doing both would be hard and not really solve anything.

You are right that the Katello workflow is still broken because there's no macro for the exact activation key, but that depends on https://github.com/Katello/katello/pull/8903.

Comment (timogoebel) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: How is the candlepin workflow different from creating the host based on e.g. puppet fact uploads?
I don't think supporting the subman workflow as a special workflow makes sense for the community.

Comment (ares) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: > but I think we should design this in a way where it does not really matter what particular tools are used to manage the host.

yes and that's what we're doing here, there may be some oddity in subscription-manager else branch, but I'll try to explain that too

> In the current implementation of the global registration template, why is there a basically unauthorized call to subscription-manager?

In this flow, subscription-manager is the way the host record is created. Currently it looks unauthorized, but the activation key is to be turned into the parameter. That is the authentication mechanism there. You're not the only one asking why we don't do the registration later in host registration template. The reason is, that would effectively cause re-registration because the host would already exist. Second reason is, subscription-manager sends us facts about the OS, so the host registration template can be already tailored to the OS, its version etc. For the generic flow, we don't have the common facter yet. So the best thing user can do is to set the OS on hostgroup level and specify hostgroup_id as a GRT parameter. We talked about improving this flow to either use what's available (facter || ohai || ansible -m) or start shipping uFacter (lzap's facter implemntation in go) or look at /etc/os-release and just parse the info from there. That all will take longer time than we have for this, so we want to have at least subscription-manager used as a facter for now.

> Doesn't this belong in the host-specific registration template that has all the required information about the host, possibly a one-time activation key tailored for that system?

The activation key shouldn't be loaded for the host, actually user should specify it. The current code is misleading, but we tried to split the work to more smaller PRs to move quicker.

> Can't we keep this simple and just create the host in the global registration template and handle the rest in the host-specific registration template?

We will for the non subscription-manager flow. That still requires one curl for GRT, second for HRT. With subscription-manager we get the limitation of it can't run the custom template as part of the registration. It also can't read the registration token. Therefore there will be more API calls in that branch.

Comment (ares) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: > Maybe having a picture visualizing the workflow would help.

yes, the documentation will defnitely need to contain that, once you have a moment, if you want please ping me on IRC, based on the past experience, short mtg saves thousands of lines :-)

Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: > How is the candlepin workflow different from creating the host based on e.g. puppet fact uploads?

With Puppet there is a server that is authenticated and authorized. The Puppet agent also sends a CSR and the Puppet CA server has a policy on that. It can be manual (default), autosign or tokens.

With subman you have the agent talking to Katello which has implemented the RHSM API. Katello then talks to Candlepin. The activation key is a static token.

The big difference is that with Puppet it can pull a catalog from the server and converge. It can use its issued client certificates to get that catalog. With subman you could do the same and implement client certificates on the HRT endpoint. Then subman wouldn't need the registration token. Not sure if it would also work with a Smart Proxy in between.

In the curl based approach you don't have any certificates so you need some authentication. The register token that was given only allows registration (and ideally would be revoked after using it once). However, it doesn't need to search for the host since the `POST` provides the token. This was suggested elsewhere as an improvement.

Comment (timogoebel) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: Ok. Thanks for the explanation. Now I believe I at least get the reasoning. So let's talk tokens or scopes.
In the current design, how are the scopes of the tokens limited?
If I get it correctly, the GRT is retrieved with permissions the current user has. But the hrt is retrieved in an anonymous scope, is it? Can I retrieve the HRT (and secrets it contains) for a host I don't own?

A disadvantage of tokens is, that they can't easily be revoked. They're made to be used several times during a specific timespan.

Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: I did suggest to change the registration to `POST /register` which would return a shell script with the HRT. That would avoid the need for the second token.

Then you would only need to deal with subman. There you can use the client certificates to retrieve HRT. The complication there is if you put a Smart Proxy in between that you then need to pass the real name. However, Smart Proxy can authenticate and we can find the Smart Proxy object. If that's found, then it can trust some parameter with the real client name.

This would completely remove the need for a Host Registration Token.

Comment (timogoebel) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: How about this flow in the GRT:

The script gathers facts about a host, in the first iteration that's just the fqdn. This can be extended by plugins. The GRT has input parameters to support other OSes, e.g. this could be a Powershell script.

(Optional/alternative part: subman register happens, ideally in the katello codebase if this requires special handling. It prefer if this were a separate PR at least)

Post call to retrieve the hrt, this passes the facts gathered in the first step. This also works when the host already exist.

The hrt is run, script cleans up and exits.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 5 above)
Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: I don't think we should include also path, we should just set base URL and path specify in the template.

1) `<%= @url %>/api/hosts` is more readable than `<%= @host_creation_url %>`
2) If user will edit template, and add call to another endpoints (let's say creating OS), they won't add new variable in controller, they will just add `<%= @url %>/my-new-api-call`
3) Chances that API path will be changed are really low
4) Too many variables : `@create_host_url`, `@register_host_url`, `@get_token_host` ... Better to have one `@url` and specify path manually.

File: app/views/unattended/provisioning_templates/registration/linux_registration_default.erb [hunk 11]
```diff
@@ -20,6 +20,23 @@ set -e
...
 <% end -%>
 
+<%= snippet 'remote_execution_ssh_keys' %>
+
+<% if host_param_true?('host_registration_insights') -%>
+if [ -f /etc/os-release ] ; then
+  . /etc/os-release
+fi
+
+if [ x$ID = xrhel ]; then
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/linux_registration_default.erb:30]: Since this is a host registration template, we already have the info in the host object. No need to perform this on the actual system.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 5 above)
Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: Adding cc @ekohl, we have same discussion here: https://github.com/theforeman/foreman/pull/7928#discussion_r487215543

Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: > Is it really worth it to have another variables just for paths, what is the chance of changing API paths?

In the case of Foreman Proxy being in between: 100%. That is not something that should be done in templates since it's not something users should even think about. URL construction is messy and error prone.

Another is that the Smart Proxy will need to be authenticated for the Katello side of it. Only then can you replace some arguments. Having all that logic in the template is IMHO a bad idea. It's part of the core functionality, not something an end user should modify.

Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: (I updated the comment meanwhile, so maybe you didn't see my changes.)

> In the case of Foreman Proxy being in between: 100%

Is it a concern for the Registration module, where we don't pass full URL?
https://github.com/theforeman/smart-proxy/pull/768/files#diff-45f33bea51f0f06fd02907550e854ae2R7
``` ruby
    def get_template(request)
      params = request.params
      params[:url] = request.env['REQUEST_URI']&.split('/register')&.first
      proxy_req = request_factory.create_get '/register', params, headers(request)
      send_request(proxy_req).body
    end
```

> Another is that the Smart Proxy will need to be authenticated for the Katello side of it. Only then can you replace some arguments.

How does it related to the Registration template?

> Having all that logic in the template is IMHO a bad idea.

What logic exactly? I agree that base URL should be defined in the controller, like this:
```ruby
@url = params[:url] || foreman_server_url // Just an example
```
but defining routes with paths should be done in the templates.

Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: > How does it related to the Registration template?

For the Katello side of it you need a Smart Proxy instance. I've explained this in other PRs, but because there's so many of them I think it's getting lost.

Another is that right now you're *developing* in this template. At some point in 2.3 we'll branch. Maybe we find a bug somewhere in the URL logic. Users may have copied 2.3 and they don't receive the bugfix. By providing the URLs in a context, the user just needs to remember to use the correct variable.

This is in line with `foreman_url`:
https://github.com/theforeman/foreman/blob/c1bc8cd6edc8b44b033ccc524989fb6c5eceacfe/lib/foreman/foreman_url_renderer.rb#L11-L36

Note that there we also properly deal with GET parameters which is always painful in templates.

That particular method may not be entirely correct for this, but that's how we have traditionally dealt with this.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 1 above)
Comment (ares) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: > So let's talk tokens or scopes.
In the current design, how are the scopes of the tokens limited?
If I get it correctly, the GRT is retrieved with permissions the current user has. But the hrt is retrieved in an anonymous scope, is it? Can I retrieve the HRT (and secrets it contains) for a host I don't own?

No, when you ask for GRT, JWT is rendered and put to it, it's then used to authenticate the HRT request.

> A disadvantage of tokens is, that they can't easily be revoked. They're made to be used several times during a specific timespan.

It will auto-expire in few hours.

> I did suggest to change the registration to POST /register which would return a shell script with the HRT. That would avoid the need for the second token.

It wouldn't. The `POST /register` also needs to be authenticated somehow. The advantage of this token is, it identifies the same user who rendered GRT, meaning we can still enforce the right org/loc and permissions when assigning additional objects. Personally I don't see a benefit of `POST /register` vs `POST /hosts`, the second guarantees all host attributes can be set as part of registration.

> Then you would only need to deal with subman. There you can use the client certificates to retrieve HRT. The complication there is if you put a Smart Proxy in between that you then need to pass the real name. However, Smart Proxy can authenticate and we can find the Smart Proxy object. If that's found, then it can trust some parameter with the real client name.

We can use smart proxy for authentication, but not for user identification and authorization. The request that is made through proxy forwards JWT, which identifies and authorizes the user. I don't think relying on proxy cert is the right way here.

> How about this flow in the GRT:

If I understand this correctly, you're suggesting combining host creation and HRT request into a single request, basically what Ewoud suggests with `POST /register`. IMHO we duplicate endpoints with that, we have one for host creation, second for HRT for specific host. API for host creation will always allow setting any host attribute, we'd need to make sure, register one will receive same updates. Perhaps we could mitigate that with shared apipie documentation, but still, seems odd if we already have endpoints for both. Also in future, we want to add more facts parsing using various facters. That will most likely create the host by different mean - facts upload. I assume we want to reuse facts upload endpoints there, hence it would again be separate request for host creation.

> Post call to retrieve the hrt, this passes the facts gathered in the first step. This also works when the host already exist.

I don't understand how your suggestion helps with existing hosts. Perhaps because it does not tell when exactly, the host is created.

Comment (timogoebel) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: > I don't understand how your suggestion helps with existing hosts. Perhaps because it does not tell when exactly, the host is created.

That actually depends. In the subman use case, it's created by subman. The post call to retrieve the HRT would only create the host if it does not exist.
Note, that there might be other use cases where the host already exists. One of them could be the VM import feature we have.

Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:65]: I've posted https://community.theforeman.org/t/rfc-simple-automatic-host-registration-wf/19588/33. Let's move the discussion back there. We've realized that at least 2 people involved in the RFC misunderstood what was proposed. I've added diagrams that I hope can help us avoid future confusion.

File: app/views/unattended/provisioning_templates/registration/linux_registration_default.erb (hunk 11 above)
Comment (ares) [reply app/views/unattended/provisioning_templates/registration/linux_registration_default.erb:30]: yeah, for a few examples of how to work with @host os information, see  e.g. https://github.com/theforeman/foreman/blob/develop/app/views/unattended/provisioning_templates/provision/kickstart_default.erb#L41-L42

HRT should fully benefit of the Foreman DB infromation

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 5 above)
Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: > For the Katello side of it you need a Smart Proxy instance. I've explained this in other PRs, but because there's so many of them I think it's getting lost.

Is it this https://github.com/Katello/katello/pull/8938 ?

I've been thinking about other cases:
- If GRT is rendered via proxy, is there a helper for generating proxy URLs like ther is for Foreman - `foreman_url`?
If not, what to do in that case?
- User is updating template and adding his own commands, for example adding facts upload via API:
```
curl -X POST <% @ foreman_url('facts_action') '%>
```
What to do if he will render the template via proxy? He can't use `foreman_url`.

**Comment by stejskalleos** [comment]
Closing, there will be more changes after the discussion in the [RFC](https://community.theforeman.org/t/rfc-simple-automatic-host-registration-wf/19588/)
